README.pod
=pod

=encoding UTF-8

=head1 NAME

Net::Async::ZMQ - IO::Async support for ZeroMQ

=head1 VERSION

version 0.002

=head1 SYNOPSIS

  use IO::Async::Loop;
  use Net::Async::ZMQ;
  use Net::Async::ZMQ::Socket;

  use ZMQ::LibZMQ3;  # or ZMQ::LibZMQ4
  use ZMQ::Constants qw(
    ZMQ_REQ
    ZMQ_NOBLOCK
    ZMQ_EVENTS ZMQ_POLLIN
  );

  my $loop = IO::Async::Loop->new;

  my $ctx = zmq_init();
  my $client_socket = zmq_socket( $ctx, ZMQ_REQ );
  zmq_connect( $client_socket, "tcp://127.0.0.1:9999" );

  my $counter = 0;

  my $zmq = Net::Async::ZMQ->new;

  zmq_sendmsg( $client_socket, "initial message" );

  $zmq->add_child(
    Net::Async::ZMQ::Socket->new(
      socket => $client_socket,
      on_read_ready => sub {
        while ( zmq_getsockopt( $client_socket, ZMQ_EVENTS ) & ZMQ_POLLIN ) {
          my $recvmsg = zmq_recvmsg( $client_socket, ZMQ_NOBLOCK );
          my $msg = zmq_msg_data($recvmsg);
          zmq_sendmsg( $client_socket, "hello @{[ $counter++ ]}" );
        }
      },
    )
  );

  $loop->add( $zmq );

  $loop->run;

=head1 DESCRIPTION

A subclass of L<IO::Async::Notifier> that can hold ZMQ sockets
that are provided by L<Net::Async::ZMQ::Socket>.

Supports sockets from the following libraries:

=over 4

=item *

L<ZMQ::LibZMQ3>

=item *

L<ZMQ::LibZMQ4>

=item *

L<ZMQ::FFI>

=back

=head1 EXTENDS

=over 4

=item * L<IO::Async::Notifier>

=back

=head1 SEE ALSO

=over 4

=item *

L<C<ZMQ_FD> in C<zmq_getsockopt>|http://api.zeromq.org/master:zmq-getsockopt>

=item *

L<ZeroMQ - Edge Triggered Notification|https://funcptr.net/2012/09/10/zeromq---edge-triggered-notification/>

=back

=head1 AUTHOR

Zakariyya Mughal <zmughal@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Zakariyya Mughal.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Changes

  Features

   - Add support for ZMQ::FFI. Fixes <https://github.com/zmughal-CPAN/p5-Net-Async-ZMQ/issues/3>.
     See PR <https://github.com/zmughal-CPAN/p5-Net-Async-ZMQ/pull/5>.
   - Update example and tests to use ZMQ_POLLIN event.
     See PR <https://github.com/zmughal-CPAN/p5-Net-Async-ZMQ/pull/7>.

0.001 2017-12-16 15:19:57-0600

  Features

   - First release.


t/lib/ReverseEcho.pm
package ReverseEcho;
# ABSTRACT: An echo server that reverses each message sent

use strict;
use warnings;

use IO::Async::Loop;
use Net::Async::ZMQ;
use Net::Async::ZMQ::Socket;

sub run {
	my ($class, $package, $n_msgs) = @_;

	my $shim;
	if( $package =~ /^ZMQ::LibZMQ[34]$/ )  {
		$shim = Shim::LibZMQx->new( $package );
	} elsif( $package =~ /^ZMQ::FFI$/ ) {
		$shim = Shim::ZMQFFI->new( $package );
	} else {
		die "Unknown package";
	}

	die "Must send at least one message from client" unless $n_msgs > 0;


	my $loop = IO::Async::Loop->new;

	my $addr = "inproc://reverse-echo";

	my $loop_done = $loop->new_future;

	my $ctx = $shim->{init}();

	my $server_socket = $shim->{socket}( $ctx, $shim->{ZMQ_REP} );
	my $client_socket = $shim->{socket}( $ctx, $shim->{ZMQ_REQ} );

	$shim->{bind}( $server_socket, $addr );
	$shim->{connect}( $client_socket, $addr );

	my $counter = 0; # message counter for client
	my @blobs; # to store messages received by the server

	# Initiate first message.
	$shim->{sendmsg}( $client_socket, "hello @{[ $counter++ ]}" );

	my $zmq = Net::Async::ZMQ->new;

	$zmq->add_child(
		Net::Async::ZMQ::Socket->new(
			socket => $client_socket,
			on_read_ready => sub {
				while ( my $recvmsg = eval{ $shim->{recvmsg}( $client_socket, $shim->{ZMQ_NOBLOCK} ) } ) {
					my $msg = $shim->{msg_data}($recvmsg);
					$shim->{sendmsg}( $client_socket, "hello @{[ $counter++ ]}" );
					if( $counter == $n_msgs + 1 ) {
						$loop_done->done;
					}
				}
			},
		)
	);

	$zmq->add_child(
		Net::Async::ZMQ::Socket->new(
			socket => $server_socket,
			on_read_ready => sub {
				while ( my $recvmsg = eval { $shim->{recvmsg}( $server_socket, $shim->{ZMQ_NOBLOCK} ) } ) {
					my $msg = $shim->{msg_data}($recvmsg);
					my $r_msg = reverse $msg;
					$shim->{sendmsg}( $server_socket, $r_msg );
					push @blobs, $msg;
				}
			},
		)
	);

	$loop->add( $zmq );

	$loop_done->on_ready(sub {
		$shim->{close}($client_socket);
		$shim->{close}($server_socket);
		$loop->stop;
	});

	$loop->loop_forever;

	\@blobs;
}

use constant SHIM_FUNCS => [ qw(
	zmq_bind
	zmq_close
	zmq_connect
	zmq_init
	zmq_msg_data
	zmq_recvmsg
	zmq_sendmsg
	zmq_socket
) ];

use constant SHIM_CONSTANTS => [ qw(
	ZMQ_REP ZMQ_REQ ZMQ_NOBLOCK
) ];

package
		Shim::LibZMQx {
	use Module::Load;

	sub new {
		my ($class, $zmq_class) = @_;
		die "Package must be ZMQ::LibZMQ[34]" unless $zmq_class =~ /^ZMQ::LibZMQ[34]$/;

		autoload $zmq_class;

		my @constants = @{ ReverseEcho::SHIM_CONSTANTS() };
		load ZMQ::Constants, @constants;

		my $self = bless {}, $class;
		$self->{_stash} = Package::Stash->new($zmq_class);

		for my $fun (@{ ReverseEcho::SHIM_FUNCS() }) {
			my $fun_no_prefix = $fun =~ s/^zmq_//r;
			$self->{$fun_no_prefix} = $self->{_stash}->get_symbol("&${fun}");
		}

		my $const_pkg = Package::Stash->new( 'ZMQ::Constants' );
		for my $constant (@constants) {
			$self->{$constant} = $const_pkg->get_symbol("&${constant}")->();
		}

		$self;
	}
}

package
		Shim::ZMQFFI {
	use Module::Load;

	sub new {
		my ($class, $zmq_class) = @_;

		die "Package must be ZMQ::FFI" unless $zmq_class =~ /^ZMQ::FFI$/;

		autoload $zmq_class;
		load 'ZMQ::FFI::Constants', @{ ReverseEcho::SHIM_CONSTANTS() };

		my $self = bless {}, $class;
		$self->{_stash} = Package::Stash->new($zmq_class);

		$self->{init} = $self->{_stash}->get_symbol('&new');
		$self->{msg_data} = sub { $_[0] };

		my %map_fun = (
			sendmsg => 'send',
			recvmsg => 'recv',
		);

		for my $fun (@{ ReverseEcho::SHIM_FUNCS() }) {
			my $fun_no_prefix = $fun =~ s/^zmq_//r;
			next if exists $self->{$fun_no_prefix};

			my $actual_fun = $fun_no_prefix;
			if( exists $map_fun{$fun_no_prefix} ) {
				$actual_fun = $map_fun{$fun_no_prefix};
			}

			$self->{$fun_no_prefix} = sub {
				my ($obj, @rest) = @_;
				$obj->can($actual_fun)->( $obj, @rest );
			}
		}

		my $const_pkg = Package::Stash->new($class);
		for my $constant (@{ ReverseEcho::SHIM_CONSTANTS() }) {
			$self->{$constant} = $const_pkg->get_symbol("&${constant}")->();
		}

		$self;
	}
}


1;


lib/Net/Async/ZMQ/Socket.pm
package Net::Async::ZMQ::Socket;
# ABSTRACT: Use a ZMQ socket asynchronously using IO::Async

use strict;
use warnings;

use Package::Stash;

use Module::Load;
use Fcntl qw(O_RDONLY);
use if $^O eq 'MSWin32', 'Win32API::File' => qw(OsFHandleOpenFd);

use base qw( IO::Async::Handle );

=method configure

  method configure( %params )

=cut
sub configure {
	my ($self, %params) = @_;

	for (qw(socket)) {
		$self->{$_} = delete $params{$_} if exists $params{$_};
	}


	if( exists $self->{socket} ) {
		my $zmq_class;
		my $zmq_fd;

		my $socket = $self->{socket};
		if( $socket->isa('ZMQ::LibZMQ3::Socket') ) {
			$self->{_zmq_class} = 'ZMQ::LibZMQ3';
			load 'ZMQ::Constants', qw(ZMQ_FD);
			$zmq_fd = ZMQ::Constants::ZMQ_FD();
		} elsif( $socket->isa('ZMQ::LibZMQ4::Socket') ) {
			$self->{_zmq_class} = 'ZMQ::LibZMQ4';
			load 'ZMQ::Constants', qw(ZMQ_FD);
			$zmq_fd = ZMQ::Constants::ZMQ_FD();
		} elsif( $socket->DOES('ZMQ::FFI::SocketRole') ) {
			$self->{_zmq_class} = 'ZMQ::FFI';
			load 'ZMQ::FFI::Constants', qw(ZMQ_FD);
			$zmq_fd = ZMQ::FFI::Constants::ZMQ_FD();
		} else {
			die "Unknown ZMQ socket: $socket";
		}

		$self->{_stash} = Package::Stash->new($self->{_zmq_class});

		$params{read_handle} = $self->_zmq_get_io_handle($self->{socket}, $zmq_fd);
	}

	$self->SUPER::configure(%params);
}

=method socket

  method socket()

Returns the underlying ZMQ socket. See the C<socket> parameter.

=cut
sub socket {
	my ($self) = @_;
	$self->{socket};
}

sub _zmq_get_io_handle {
	my ($self, $socket, $zmq_fd) = @_;

	my $zmq_getsockopt =
		$self->{_zmq_class} eq 'ZMQ::FFI'
		? sub { my ($socket, $opt) = @_; $socket->get( $opt, 'int' ) }
		: $self->{_stash}->get_symbol('&zmq_getsockopt');
	my $zmq_getsockopt_uint64 =
		$self->{_zmq_class} eq 'ZMQ::FFI'
		? sub { my ($socket, $opt) = @_; $socket->get( $opt, 'uint64_t' ) }
		: $self->{_stash}->get_symbol('&zmq_getsockopt_uint64');

	my $fd;

	if( $^O eq 'MSWin32' ) {
		# `SOCKET` data type is a `uint64` on Windows x64.
		my $socket_handle = $zmq_getsockopt_uint64->( $socket, $zmq_fd );
		# Converts OS socket handle to a C runtime file descriptor.
		$fd = OsFHandleOpenFd($socket_handle, O_RDONLY);
	} else {
		$fd = $zmq_getsockopt->( $socket, $zmq_fd );
	}

	# Use dup() on the ZMQ file descriptor so that Perl can close the
	# handle without closing the ZMQ handle.
	#
	# Also, avoid using IO::Handle here due to how it closes the file
	# handle automatically.
	open(my $io_handle, "<&", $fd);

	$io_handle;
}

1;
__END__
=head1 DESCRIPTION

A subclass of L<IO::Async::Handle> that works with ZMQ sockets.

It currently handles sockets from L<ZMQ::LibZMQ3>, L<ZMQ::LibZMQ4>, and
L<ZMQ::FFI> on both Unix-like systems and Windows.

=head1 PARAMETERS

=head2 socket

A ZMQ socket.

Takes the type

  InstanceOf['ZMQ::LibZMQ3::Socket']
  | InstanceOf['ZMQ::LibZMQ4::Socket']
  | ConsumerOf['ZMQ::FFI::SocketRole']


lib/Net/Async/ZMQ.pm
package Net::Async::ZMQ;
# ABSTRACT: IO::Async support for ZeroMQ

use strict;
use warnings;

use base qw( IO::Async::Notifier );

1;
__END__
=head1 SYNOPSIS

  use IO::Async::Loop;
  use Net::Async::ZMQ;
  use Net::Async::ZMQ::Socket;

  use ZMQ::LibZMQ3;  # or ZMQ::LibZMQ4
  use ZMQ::Constants qw(
    ZMQ_REQ
    ZMQ_NOBLOCK
    ZMQ_EVENTS ZMQ_POLLIN
  );

  my $loop = IO::Async::Loop->new;

  my $ctx = zmq_init();
  my $client_socket = zmq_socket( $ctx, ZMQ_REQ );
  zmq_connect( $client_socket, "tcp://127.0.0.1:9999" );

  my $counter = 0;

  my $zmq = Net::Async::ZMQ->new;

  zmq_sendmsg( $client_socket, "initial message" );

  $zmq->add_child(
    Net::Async::ZMQ::Socket->new(
      socket => $client_socket,
      on_read_ready => sub {
        while ( zmq_getsockopt( $client_socket, ZMQ_EVENTS ) & ZMQ_POLLIN ) {
          my $recvmsg = zmq_recvmsg( $client_socket, ZMQ_NOBLOCK );
          my $msg = zmq_msg_data($recvmsg);
          zmq_sendmsg( $client_socket, "hello @{[ $counter++ ]}" );
        }
      },
    )
  );

  $loop->add( $zmq );

  $loop->run;

=head1 DESCRIPTION

A subclass of L<IO::Async::Notifier> that can hold ZMQ sockets
that are provided by L<Net::Async::ZMQ::Socket>.

Supports sockets from the following libraries:

=begin :list

* L<ZMQ::LibZMQ3>

* L<ZMQ::LibZMQ4>

* L<ZMQ::FFI>

=end :list

=head1 SEE ALSO

=begin :list

* L<C<ZMQ_FD> in C<zmq_getsockopt>|http://api.zeromq.org/master:zmq-getsockopt>

* L<ZeroMQ - Edge Triggered Notification|https://funcptr.net/2012/09/10/zeromq---edge-triggered-notification/>

=end :list

=cut
